### ISSUE TITLE
I have some ideas, opinions?

### ISSUE THREAD

**Issue by tecosaur** [James-Yu/LaTeX-Workshop#844]
## I've had some ideas, I'd appreciate feedback on them
_i.e. are they worth pursuing?_

### Math Wrap Keybinding
 - I've been trying out the shortcut `ctrl+m ctrl+w` for _'wrap with math_ i.e. `\( … \)` 
 - have found it very handy so far
 - not sure about a corresponding snippet - perhaps `@mw`??
 - Can implement myself

### Section promotion / demotion
 - Idea: ability to promote and demote section commands
    - i.e. `chapter` ↔ `section` ↔ `subsection` ↔ `subsubsection` ↔ `paragraph` ↔ `subparagraph`
 - Idea came about because of how freqently I change the 'priority' of various parts of the documents I work on
 - I'm currently thinking `ctrl+l ctrl+p` (for promote) and `ctrl+l ctrl+d` (for demote) may not be a bad idea
 - May be able to implement myself

### Auto-rename matching tag
 - Basically [this](https://marketplace.visualstudio.com/items?itemName=formulahendry.auto-rename-tag) but for LaTeX
 - Probably wouldn't be able to implement myself without help

**Comment by jlelong** [comment]
As I am not a fan of keybindings, I will let others respond. Just a small comment: math Wrapping already exists. Select text, hit `\`, the first entry in the list does what you want.

If what you mean by tag renaming is changing the name in a `\begin{...} / \end{...}` accordingly. This can be easily achieved by using _LaTeX Workshop: Select the current environment name_ and enter the new name.

**Comment by tecosaur** [comment]
Selecting text then hitting `\` just overwrites the text for me - I feel like I'm missing something.

Re: environment setting - Never noticed 😅 thanks for the tip (I should really read through the commands list before having ideas like these)

One another note, regarding keybindings vs. snippets, the thing I have about snippets, particularly within maths is that (at least without me working out some keybinding modifications) I can't do snippet-in-snippet. 
I.e. Pressing tab after typing `@a` within a `\frac` from `@/` takes me to the second part of the fraction, and doesn't expand `@a`  —  do you know any way around this? It's why I prefer keybindings at the moment.

**Comment by jlelong** [comment]
It does not seem possible to nest snippets when the first one contains placeholders. We have already come across that, see #609.

**Comment by jlelong** [comment]
> Selecting text then hitting `\` just overwrites the text for me - I feel like I'm missing something.

Check that `latex-workshop.intellisense.surroundCommand.enabled` is set to `true`, default is `false`. I have just added this to the README file.

**Comment by Astrantia** [comment]
@jlelong any reason we don't mimic Emmet's behavior with surrounding/wrapping? Right now when we click `\` our highlighted text is deleted whereas in Emmet, when you click the wrap keybinding the highlighted text is not deleted and makes up for better User Experience.

**Comment by jlelong** [comment]
Actually, I was wondering the same earlier today. I have started working on this but I have been struggling to define `\` as a shortcut. This would fix #517.

**Comment by Astrantia** [comment]
@jlelong I don't see why we need to define the shortcut as `\`. I don't think there are too many benefits doing that. Let's just define it as `(CTRL+L) + (CTRL+W)` for L(atex) W(rap).

**Comment by jlelong** [comment]
Because I would like to keep the current keybinding but change the way it is actually handled. Let you know more if I can my way through keyboard layouts...

**Comment by jlelong** [comment]
Based on [this](https://code.visualstudio.com/docs/getstarted/keybindings#_keyboard-layouts), there is no hope to use `\` reliably as a keybinding on all keyboard layouts. Therefore, my idea is
1. To keep the current surround feature with `\` but clearly mark it as deprecated with the idea of eventually removing it.
1. To introduce a new surround feature designed as in Emmet using a more standard keybinding: this would fix #517 and would not require to keep track of all selection changes anymore.
@James-Yu what do you think of it?

**Comment by James-Yu** [comment]
It's an absolutely welcoming change, though I'm looking into the auto-completion for `\input`. Will turn to that afterwards if no progress were made.

**Comment by jlelong** [comment]
I have just pushed the new surround feature. @James-Yu I do not know what is the best way of marking the old use of `\` deprecated (as of now it is mentioned in [README.md](https://github.com/James-Yu/LaTeX-Workshop/blob/master/README.md#L35)).

I am closing this issue and suggest to start a new one for new ideas/features. It is easier to keep track of history when one issue corresponds to one feature.
